bin/chef-cluster

```diff
@@ -0,0 +1,115 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+cmd=$1
+workercount=${2:-auto}
+vhostconfig=/etc/apache2/vhosts.d/chef-cluster.conf
+serverconfig=/etc/sysconfig/chef-server
+webroot=/srv/www/chef-cluster
+
+baseport=4001
+
+balancerstart=BEGIN-BALANCER-MEMBER
+balancerend=END-BALANCER-MEMBER
+
+usage() {
+  echo "usage: chef-cluster <enable|disable> [<workercount>|auto]"
+  exit 1
+}
+
+case $cmd in
+  enable)
+    [ "$workercount" == "auto" ] && workercount=$(grep -c ^processor /proc/cpuinfo)
+    [ "$workercount" -ge 2 ] || { echo "ERROR: workercount must be >= 2"; exit 2; }
+
+    # apache configuration
+    mkdir -p "$webroot"
+    a2enmod -q lbmethod_byrequests || a2enmod lbmethod_byrequests
+    test -e "$vhostconfig" || cat <<EOF > $vhostconfig
+# configuration generated by: chef-cluster script
+Listen 4000
+
+<VirtualHost *:4000>
+  DocumentRoot /srv/www/chef-cluster
+  ErrorLog /var/log/apache2/chef-cluster-error.log
+  CustomLog /var/log/apache2/chef-cluster-access.log combined
+
+  <Location />
+    Require all granted
+  </Location>
+
+  <Proxy balancer://chefworkers>
+    # $balancerstart
+    # $balancerend
+  </Proxy>
+
+  <Location /balancer-manager>
+    SetHandler balancer-manager
+    Require host localhost
+    Require host 127.0.0.1
+  </Location>
+
+  RewriteEngine On
+  RewriteCond %{REQUEST_URI} !=/balancer-manager
+  RewriteCond %{DOCUMENT_ROOT}/%{REQUEST_FILENAME} !-f
+  RewriteRule ^/(.*)$ balancer://chefworkers%{REQUEST_URI} [P,QSA,L]
+</VirtualHost>
+EOF
+
+    awk -i inplace "
+      /$balancerstart/ { print; f=1} !f;
+      /$balancerend/ {
+        for (i=$baseport;i<$baseport+$workercount;i++) {
+          print \"    BalancerMember http://127.0.0.1:\"i
+        };
+        print;
+        f=0
+      }
+    " "$vhostconfig"
+    echo "$vhostconfig created/updated"
+
+    # chef-server configuration
+    test -e "$serverconfig" && mv "$serverconfig" "$serverconfig.old"
+    cat <<EOF > $serverconfig
+# configuration generated by: chef-cluster script
+PORT=$baseport
+OPTIONS="-N -c $workercount"
+EOF
+    echo "$serverconfig created/updated "
+
+    systemctl restart chef-server
+    systemctl reload apache2
+    echo "services restarted"
+
+    echo "chef cluster enabled with $workercount workers."
+  ;;
+
+  disable)
+    mv "$vhostconfig" "$vhostconfig.disabled" 2> /dev/null || { echo "ERROR: disabling $vhostconfig failed. Is chef cluster enabled?"; exit 3; }
+    echo "$vhostconfig disabled"
+    mv "$serverconfig" "$serverconfig.disabled" 2> /dev/null || { echo "ERROR: disabling $serverconfig failed. Is chef cluster enabled?"; exit 4; }
+    echo "$serverconfig disabled"
+    systemctl reload apache2
+    systemctl restart chef-server
+    echo "services restarted"
+    echo "chef cluster disabled."
```

Comment:
I would accept calling this while it's already disabled, and only restart/reload when needed. I'm also not sure what's the benefit of moving the files with a `.disabled` prefix, since the script can recreate them.

```shell
if test -f "$vhostconfig"; then
  rm -f "$vhostconfig"
  systemctl reload apache2
fi
if test -f "$serverconfig"; then
  rm -f "$serverconfig"
  systemctl restart chef-server
fi
echo "chef-server cluster disabled."
```

Reply:
The benefit is that these files can be modified by the user. They are not managed by chef and can exist before we enable the cluster and/or be modified while the cluster is running.